nemo/collections/nlp/data/language_modeling/megatron/indexed_retrieval_dataset.py

```diff
@@ -0,0 +1,471 @@
+                self._bin_buffer,
+                dtype=np.int64,
+                count=self._len,
+                offset=offset + self._sizes.nbytes + self._pointers.nbytes,
+            )
+            logging.info("    reading chunk address...")
+            self._chunk_address = np.frombuffer(
+                self._bin_buffer,
+                dtype=np.int64,
+                count=self.num_chunks,
+                offset=offset + self._sizes.nbytes + self._pointers.nbytes + self._chunk_id_start.nbytes,
+            )
+
+        def get_chunk_address(self, chunk_id):
+            """ get the chunk address from chunk id
+            """
+            return self._chunk_address[chunk_id]
+
+        def get_chunk_id(self, sentence_id, position):
+            """ get the chunk id from sentence idx and offset position.
+            """
+            return (self._chunk_id_start[sentence_id] + position // self.chunk_size).item()
+
+        def __del__(self):
+            self._bin_buffer_mmap._mmap.close()
+            del self._bin_buffer_mmap
+
+        @property
+        def dtype(self):
+            return self._dtype
+
+        @property
+        def sizes(self):
+            return self._sizes
+
+        @lru_cache(maxsize=8)
+        def get_chunk(self, chunk_id):
+            return self._pointers[i], self._sizes[i]
+
+        @lru_cache(maxsize=8)
+        def __getitem__(self, i):
+            return self._pointers[i], self._sizes[i]
+
+        def __len__(self):
+            return self._len
+
+    def __init__(self, path, skip_warmup=False):
+        super().__init__()
+
+        self._path = None
+        self._index = None
+        self._bin_buffer = None
+
+        self._do_init(path, skip_warmup)
+
+    def __getstate__(self):
+        return self._path
+
+    # def __setstate__(self, state):
+    #     self._do_init(state)
+
+    def _do_init(self, path, skip_warmup):
+        self._path = path
+        self._index = self.Index(index_file_path(self._path), skip_warmup)
+
+        if not skip_warmup:
+            logging.info("    warming up data mmap file...")
+            _warmup_mmap_file(data_file_path(self._path))
+        logging.info("    creating numpy buffer of mmap...")
+        self._bin_buffer_mmap = np.memmap(data_file_path(self._path), mode='r', order='C')
+        logging.info("    creating memory view of numpy buffer...")
+        self._bin_buffer = memoryview(self._bin_buffer_mmap)
+
+    def __del__(self):
+        self._bin_buffer_mmap._mmap.close()
+        del self._bin_buffer_mmap
+        del self._index
+
+    def __len__(self):
+        return len(self._index)
+
+    # @lru_cache(maxsize=8)
+    def __getitem__(self, idx):
+        if isinstance(idx, int):
+            ptr, size = self._index[idx]
+            np_array = np.frombuffer(self._bin_buffer, dtype=self._index.dtype, count=size, offset=ptr)
+            return np_array
+        elif isinstance(idx, slice):
```

Comment:
Cool. Did you happen to notice improvement in speed for fetching consecutive items?